### DISCUSSION TITLE
Preserve page when clicking Cancel on Edit page

### DISCUSSION THREAD

**Discussion by hrvoj3e** [filamentphp/filament#4079] [Help] [answered]
How would I redirect with filters,page,size included when clicking on Cancel ?

Steps to reproduce
- go to resource list
- go to page 5
- edit record
- click cancel
- it goes to list on page=1 insted of page 5
- is this a bug?

Can be reproduced on resource Posts
 - https://demo.filamentphp.com/blog/posts

**Comment by studiowizjo** [comment]
Some workaround of this problem can be placing this method in your `EditRecord` file:

```
protected function getCancelFormAction(): Action
{
    return Action::make('cancel')
        ->label(__('filament::resources/pages/edit-record.form.actions.cancel.label'))
        ->extraAttributes(['onclick' => 'history.back()'])
        ->color('secondary');
}
```

I said "workaround" because is works the same as previous button in browser, so downside of this is that when you visit edit page directly, the button will not work, as there is no page to go back.

**Reply by hrvoj3e** [reply]
Nice to know about `onclick` with history :+1: 

But it seems to me that this is bad UX. 
User can order by column, use filters and one Cancel would reset results on list.

**Comment by danharrin** [comment, marked as answer]
Definitely not a bug.

You need to use the session for this.

Page size is persisted by default I think.

Persisting search in session: https://filamentphp.com/docs/2.x/tables/columns#persist-search-in-session

Persisting filters in session: https://filamentphp.com/docs/2.x/tables/filters#persist-filters-in-session

**Both of these methods need to be put on the List page class, not the Resource class.**

**Reply by RX-77IT** [reply]
Sorry for butting in, but is there some function that can persist the page number?
Page size, filters and search already work wonderfully (thank you BTW).

**Reply by danharrin** [reply]
Not yet, but you could PR that!

**Reply by hrvoj3e** [reply]
Too bad for page. Is page persist on todo feature list maybe?

**Reply by danharrin** [reply]
Its not a feature I would use, you can PR it yourself though, this is open source.

**Comment by flyingeek** [comment]
Here is my suggestion crossposted from #4364 

I'm not good enough to propose a PR but:

Create the trait below then add to your EditPage or CreatePage the trait

```php
<?php

namespace App\Traits;

use Filament\Pages\Actions\Action;
use Filament\Resources\Pages\Page;

trait HasBackUrl
{
    public ?string $backUrl = null;

    public function mount($record=null): void
    {
        $this->backUrl = url()->previous();
        parent::mount($record);
    }

    protected function getRedirectUrl(): string
    {
        return $this->backUrl;
    }

    protected function getCancelFormAction(): Action
    {
        return parent::getCancelFormAction()
            ->url(fn(Page $livewire) => $livewire->backUrl ?? static::getResource()::getUrl());
    }
}
```

Then in your EditUser page (not the resource) add for exemple:

```php
use Filament\Resources\Pages\EditRecord;
use App\Traits\HasBackUrl;

class EditUser extends EditRecord
{
    use HasBackUrl;   // <- add this line
   //....
}
```
